File: src/services/audit/detectors/ad/permissions/moderate/force-change-password.ts
/**
 * ACL_FORCECHANGEPASSWORD - ForceChangePassword extended right
 */

import { AclEntry } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { getUniqueObjects } from '../utils';

/**
 * Check for ForceChangePassword extended right
 */
export function detectAclForceChangePassword(aclEntries: AclEntry[], includeDetails: boolean): Finding {
  const FORCE_CHANGE_PASSWORD_GUID = '00299570-246d-11d0-a768-00aa006e0529';

  const affected = aclEntries.filter((ace) => {
    return ace.objectType === FORCE_CHANGE_PASSWORD_GUID;
  });

  const uniqueObjects = getUniqueObjects(affected);
  const totalInstances = affected.length;

  return {
    type: 'ACL_FORCECHANGEPASSWORD',
    severity: 'medium',
    category: 'permissions',
    title: 'ACL Force Change Password',
    description: 'ExtendedRight to force password change on user accounts. Can reset passwords without knowing current password.',
    count: uniqueObjects.length,
    totalInstances: totalInstances !== uniqueObjects.length ? totalInstances : undefined,
    affectedEntities: includeDetails ? uniqueObjects : undefined,
  };
}

/**
 * Detect User-Force-Change-Password rights
 *
 * This right allows resetting user passwords without knowing the current password.
 *
 * @param aclEntries - Array of ACL entries
 * @param includeDetails - Whether to include affected entity details
 * @returns Finding for ACL_USER_FORCE_CHANGE_PASSWORD
 */
export function detectAclUserForceChangePassword(
  aclEntries: AclEntry[],
  includeDetails: boolean
): Finding {
  // User-Force-Change-Password: 00299570-246d-11d0-a768-00aa006e0529
  const FORCE_CHANGE_PASSWORD_GUID = '00299570-246d-11d0-a768-00aa006e0529';

  const affected = aclEntries.filter((ace) => {
    return ace.objectType?.toLowerCase() === FORCE_CHANGE_PASSWORD_GUID;
  });

  const uniqueObjects = getUniqueObjects(affected);
  const totalInstances = affected.length;

  return {
    type: 'ACL_USER_FORCE_CHANGE_PASSWORD',
    severity: 'medium',
    category: 'permissions',
    title: 'User-Force-Change-Password Rights',
    description:
      'Principals with rights to force password change on user accounts. ' +
      'Can reset passwords to take over accounts.',
    count: uniqueObjects.length,
    totalInstances: totalInstances !== uniqueObjects.length ? totalInstances : undefined,
    affectedEntities: includeDetails ? uniqueObjects : undefined,
  };
}

File: internal/audit/detectors/ad/permissions/moderate/force-change-password.go
package moderate

import (
	"context"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// ForceChangePasswordDetector detects ForceChangePassword extended right
type ForceChangePasswordDetector struct {
	audit.BaseDetector
}

// NewForceChangePasswordDetector creates a new detector
func NewForceChangePasswordDetector() *ForceChangePasswordDetector {
	return &ForceChangePasswordDetector{
		BaseDetector: audit.NewBaseDetector("ACL_FORCECHANGEPASSWORD", audit.CategoryPermissions),
	}
}

// Detect executes the detection
func (d *ForceChangePasswordDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	const forceChangePasswordGUID = "00299570-246d-11d0-a768-00aa006e0529"

	var affected []types.ACLEntry

	for _, ace := range data.ACLEntries {
		if strings.ToLower(ace.ObjectType) == forceChangePasswordGUID {
			affected = append(affected, ace)
		}
	}

	uniqueObjects := helpers.GetUniqueObjects(affected)
	totalInstances := len(affected)

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "ACL Force Change Password",
		Description: "ExtendedRight to force password change on user accounts. Can reset passwords without knowing current password.",
		Count:       len(uniqueObjects),
	}

	if totalInstances != len(uniqueObjects) {
		finding.TotalInstances = totalInstances
	}

	if data.IncludeDetails && len(uniqueObjects) > 0 {
		entities := make([]types.AffectedEntity, len(uniqueObjects))
		for i, dn := range uniqueObjects {
			entities[i] = types.AffectedEntity{
				Type: "user",
				DN:   dn,
			}
		}
		finding.AffectedEntities = entities
	}

	return []types.Finding{finding}
}

// UserForceChangePasswordDetector detects User-Force-Change-Password rights
type UserForceChangePasswordDetector struct {
	audit.BaseDetector
}

// NewUserForceChangePasswordDetector creates a new detector
func NewUserForceChangePasswordDetector() *UserForceChangePasswordDetector {
	return &UserForceChangePasswordDetector{
		BaseDetector: audit.NewBaseDetector("ACL_USER_FORCE_CHANGE_PASSWORD", audit.CategoryPermissions),
	}
}

// Detect executes the detection
func (d *UserForceChangePasswordDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// User-Force-Change-Password: 00299570-246d-11d0-a768-00aa006e0529
	const forceChangePasswordGUID = "00299570-246d-11d0-a768-00aa006e0529"

	var affected []types.ACLEntry

	for _, ace := range data.ACLEntries {
		if strings.ToLower(ace.ObjectType) == forceChangePasswordGUID {
			affected = append(affected, ace)
		}
	}

	uniqueObjects := helpers.GetUniqueObjects(affected)
	totalInstances := len(affected)

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "User-Force-Change-Password Rights",
		Description: "Principals with rights to force password change on user accounts. Can reset passwords to take over accounts.",
		Count:       len(uniqueObjects),
	}

	if totalInstances != len(uniqueObjects) {
		finding.TotalInstances = totalInstances
	}

	if data.IncludeDetails && len(uniqueObjects) > 0 {
		entities := make([]types.AffectedEntity, len(uniqueObjects))
		for i, dn := range uniqueObjects {
			entities[i] = types.AffectedEntity{
				Type: "user",
				DN:   dn,
			}
		}
		finding.AffectedEntities = entities
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewForceChangePasswordDetector())
	audit.MustRegister(NewUserForceChangePasswordDetector())
}
